### PR TITLE
Packet 3.3 - Fall Packet MVP

### PR DESCRIPTION
**Pull request by devinmatte**
New Features:
- Push Notifications
- Automated Emails
- Speed Improvements

### CODE REVIEW

File: config.env.py [hunk 1]
```diff
@@ -31,5 +32,23 @@
 LDAP_BIND_DN = environ.get("PACKET_LDAP_BIND_DN", None)
 LDAP_BIND_PASS = environ.get("PACKET_LDAP_BIND_PASS", None)
 
+# Mail Config
+MAIL_PROD = strtobool(environ.get("PACKET_MAIL_PROD", 'False'))
+MAIL_SERVER = environ.get("PACKET_MAIL_SERVER", "thoth.csh.rit.edu")
+MAIL_USERNAME = environ.get("PACKET_MAIL_USERNAME", "packet@csh.rit.edu")
+MAIL_PASSWORD = environ.get("PACKET_MAIL_PASSWORD", None)
+MAIL_USE_TLS = strtobool(environ.get("PACKET_MAIL_TLS", 'True'))
```
Comment (JoelEager) [inline config.env.py]: There's inconstant use of `'` and `"` in this file. Not a big deal but may want to fix it.

File: packet/mail.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
+from flask import render_template
+from flask_mail import Mail, Message
+
+from packet import app
+
+mail = Mail(app)
+
+
+def send_start_packet_mail(packet):
+    if app.config['MAIL_PROD']:
```
Comment (JoelEager) [inline packet/mail.py]: This file has quote issues too.

File: packet/notifications.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+            app.logger.info("The notification ({}) sent out successfully".format(notification.post_body))
+        else:
+            app.logger.warn("The notification ({}) was unsuccessful".format(notification.post_body))
+
+
+def packet_signed_notification(packet, signer):
+    subscriptions = NotificationSubscription.query.filter_by(freshman_username=packet.freshman_username)
+    if subscriptions:
+        notification_body = post_body
+        notification_body["contents"]["en"] = signer + ' signed your packet! Congrats or I\'m Sorry'
```
Comment (JoelEager) [inline packet/notifications.py]: And same here

File: packet/routes/api.py [hunk 4]
```diff
@@ -20,25 +24,53 @@ def sign(packet_id, info):
...
             db.session.add(MiscSignature(packet=packet, member=info["uid"]))
             app.logger.info("Member {} signed packet {} as a misc".format(info["uid"], packet_id))
+            packet_signed_notification(packet, info["uid"])
             return commit_sig(packet, was_100)
         else:
             # Check if the freshman is onfloor and if so, sign that row
             for sig in filter(lambda sig: sig.freshman_username == info["uid"], packet.fresh_signatures):
                 sig.signed = True
                 app.logger.info("Freshman {} signed packet {}".format(info["uid"], packet_id))
+                packet_signed_notification(packet, info["uid"])
```
Comment (JoelEager) [inline packet/routes/api.py]: Maybe this should be moved to the `commit_sig()` function? That'd reduce the code duplication going on here. And could you move that function above this one in the file? That'd make things a bit more readable.

File: packet/routes/api.py [hunk 5]
```diff
@@ -20,25 +24,53 @@ def sign(packet_id, info):
...
     return "Error: Signature not valid.  Reason: Unknown"
 
+
+@app.route("/api/v1/subscribe/", methods=["POST"])
+@packet_auth
+@before_request
+def subscribe(info):
+    data = request.form
+    if app.config["REALM"] == "csh":
+        subscription = NotificationSubscription(token=data['token'], member=info["uid"])
```
Comment (JoelEager) [inline packet/routes/api.py]: _Joel is a broken record._

File: packet/routes/shared.py [hunk 6]
```diff
@@ -68,3 +68,15 @@ def packets(info=None):
     open_packets.sort(key=packet_sort_key, reverse=True)
 
     return render_template("active_packets.html", info=info, packets=open_packets)
+
+
+@app.route('/sw.js', methods=['GET'])
```
Comment (JoelEager) [inline packet/routes/shared.py]: Same story; new file.

File: packet/routes/api.py (hunk 5 above)
Comment (devinmatte) [reply packet/routes/api.py]: What do you mean?

File: packet/routes/shared.py (hunk 6 above)
Comment (devinmatte) [reply packet/routes/shared.py]: Why new file? It's shared?

File: packet/routes/api.py (hunk 5 above)
Comment (JoelEager) [reply packet/routes/api.py]: I was talking about the quotes in this file; sorry.

**Pull request merged**